[Issue#74] Expand location template/macro if found in copts

ravimandala opened the pull request:
Bazel supports location template and provides [an API](https://github.com/bazelbuild/bazel/blob/52446a12b7435e850347dda144566280625cf498/src/main/java/com/google/devtools/build/lib/skylarkbuildapi/SkylarkRuleContextApi.java#L715-L747) to expand location, with the path of the corresponding output file, if and when the `context` is accessible.

Tulsi currently doesn't support the location template and messes up the paths/ flags by splitting the template by space.
Refer to https://github.com/bazelbuild/tulsi/issues/74 for more details.

\cc @DavidGoldman @reinhillmann @dierksen @googlebot 

Change-Id: I397590c93a05918206fe9d78867095336d6532ae

ravimandala committed:
[Issue#74] Expand location template/macro if found in copts

Bazel supports location template and provides [an API](https://github.com/bazelbuild/bazel/blob/52446a12b7435e850347dda144566280625cf498/src/main/java/com/google/devtools/build/lib/skylarkbuildapi/SkylarkRuleContextApi.java#L715-L747) to expand location, with the path of the corresponding output file, if and when the `context` is accessible.

Tulsi currently doesn't support the location template and messes up the paths/ flags by splitting the template by space.
Refer to https://github.com/bazelbuild/tulsi/issues/74 for more details.

Change-Id: I397590c93a05918206fe9d78867095336d6532ae

ravimandala commented:
Also created this - https://bazel-review.googlesource.com/c/tulsi/+/89190

ravimandala commented:
Thanks to @ob for chipping in for this change. 

@DavidGoldman @dierksen Can you please take a look at this? 
\cc @reinhillmann

dierksen commented:
Posted this to #74, when I meant to post it here, but anyway, can you provide an example that we could add for an integration test?

dierksen commented on src/TulsiGenerator/Bazel/tulsi/tulsi_aspects.bzl:
```diff
@@ -710,6 +710,16 @@ def _tulsi_sources_aspect(target, ctx):
+    if copts_attr:
+        expanded_attrs = []
+        for attr in copts_attr:
+            if '$(location' in attr:
```
Does calling `ctx.expand_location` do anything wrong if it's used on something that _doesn't_ have a `$(location <target>)` template?

i.e. could you get rid of the `if` block and just do something like:

```
if copts_attr:
    copts_attr = [ctx.expand_location(attr, [target]) for attr in copts_attr]
```

It seems like that should work (existing tests all pass with that at least), but having a good example to use for an integration test would be helpful.

ravimandala replied:
Simplified the block as per your suggestion. That is working as expected.

ravimandala commented:
@dierksen I simplified the location expansion logic as per your suggestion. Also modified the `.gitignore` to exclude Tulsi's own Xcode project files and tulsi-workspace symlink. Please take a look.

dierksen commented on .gitignore:
```diff
@@ -4,3 +4,10 @@
+tulsi-*
+
+# Xcode project
+/Tulsi/
```
Should this be `/Tulsi.xcodeproj/`?

dierksen replied:
Also, I'm going to hand off any further reviewing to @DavidGoldman or one of the other owners.

ravimandala replied:
Yes, that makes sense. Changed it to just `/Tulsi.xcodeproj/`

DavidGoldman commented on src/TulsiGenerator/Bazel/tulsi/tulsi_aspects.bzl:
```diff
@@ -710,6 +710,9 @@ def _tulsi_sources_aspect(target, ctx):
                         _collect_bundle_imports(rule_attr))
 
     copts_attr = _get_opt_attr(rule_attr, "copts")
+    if copts_attr:
```
Please add a comment here mentioning why the expansion is needed (an example would be good)

ravimandala replied:
I added a detailed comment with an example. 
@ob has an example in https://github.com/bazelbuild/bazel/issues/7083 in the form of a `repro.sh` to reproduce this issue. Does that help for integration testing?

DavidGoldman commented:
Friendly ping

ob commented:
@DavidGoldman I think @ravimandala is waiting for https://github.com/bazelbuild/bazel/issues/7083 to be resolved before adding this change as it can potentially break valid builds if people add an actual call to `$(location)`...

ob commented:
https://github.com/bazelbuild/bazel/issues/7083 is fixed. @ravimandala can you add the comment/example that @DavidGoldman requested?

ravimandala commented:
@DavidGoldman, @ob has an example in https://github.com/bazelbuild/bazel/issues/7083 in the form of a `repro.sh` to reproduce this issue. Does that help with integration test?

ravimandala commented:
@DavidGoldman, Can you please take a look?

ravimandala commented:
@dierksen, @DavidGoldman - Gentle reminder.

DavidGoldman commented:
> @DavidGoldman, @ob has an example in [bazelbuild/bazel#7083](https://github.com/bazelbuild/bazel/issues/7083) in the form of a `repro.sh` to reproduce this issue. Does that help with integration test?

It would you be good to modify the [AspectTests](https://github.com/bazelbuild/tulsi/blob/master/src/TulsiGeneratorIntegrationTests/AspectTests.swift) to check that this works as intended (you can add some new copts to an existing target).

ravimandala commented:
Ahh, now I see what you mean. Let me add a test for this feature.

thii commented:
@ravimandala Friendly ping—any update? :)

keith commented:
Anyone know the state of this one?

keith commented:
Related: https://github.com/bazelbuild/tulsi/pull/157

keith commented:
Thanks for submitting this PR but this repository is being deprecated. Please switch to [rules_xcodeproj](https://github.com/buildbuddy-io/rules_xcodeproj) as a more complete and maintained alternative.
